File: src/main/java/org/dukcode/ps/codetree/trail02/chapter02/lesson02/TestLeastCommonMultipleUsingRecursiveFunction.java
package org.dukcode.ps.codetree.trail02.chapter02.lesson02;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/test-least-common-multiple-using-recursive-function/description">LINK</a>
 */
public class TestLeastCommonMultipleUsingRecursiveFunction {

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    int n = Integer.parseInt(br.readLine());

    int[] arr = new int[n];
    st = new StringTokenizer(br.readLine());
    for (int i = 0; i < n; i++) {
      arr[i] = Integer.parseInt(st.nextToken());
    }

    bw.write(String.valueOf(lcm(arr, 0, n)));

    br.close();
    bw.close();
  }

  private static int lcm(int[] arr, int frInIdx, int toExIdx) {
    if (frInIdx + 1 == toExIdx) {
      return arr[frInIdx];
    }

    int a = arr[frInIdx];
    int b = lcm(arr, frInIdx + 1, toExIdx);
    int gcd = 0;
    for (int i = Math.min(a, b); i > 0; --i) {
      if (a % i == 0 && b % i == 0) {
        gcd = i;
        break;
      }
    }

    return a * b / gcd;
  }
}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter02/lesson02/TestLeastCommonMultipleUsingRecursiveFunction.kt
package org.dukcode.ps.codetree.trail02.chapter02.lesson02.testLeastCommonMultipleUsingRecursiveFunction

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/test-least-common-multiple-using-recursive-function/description">LINK</a>
 */
fun main() {
    skipLine()
    val arr = parseInts()

    bw.write(arr.lcm().toString())

    bw.flush()
}

private fun IntArray.lcm() = this.lcm(0, size)

private fun IntArray.lcm(
    frIn: Int,
    toEx: Int,
): Int {
    if (toEx - frIn == 1) {
        return this[frIn]
    }

    val next = this.lcm(frIn + 1, toEx)

    return this[frIn] * next / gcd(this[frIn], next)
}

private fun gcd(
    a: Int,
    b: Int,
): Int {
    if (b == 0) {
        return a
    }

    if (a < b) {
        return gcd(b, a)
    }

    return gcd(b, a % b)
}

private fun parseInts() =
    StringTokenizer(br.readLine()).run {
        IntArray(countTokens()) {
            nextToken().toInt()
        }
    }

private fun skipLine() = br.readLine()
